add AssetTemplate

stefansc1 opened the pull request:
implements #283

Add new "AssetTemplate" model with page in Django admin.
Add "template" section to asset modal, collapsed by default.
In this section, a user can save a template for a) just this project or b) their account. In addition, they may request to make it available to all users. Users can give their templates a name and an optional description. For admin purposes, asset type, creation date, associated project and user are also saved. Asset data is saved as JSON in DB.
A user may also load an existing template (if one exists with a matching asset type and visibility), filling out all asset form fields. The asset form must still be saved manually (in case the user wants to make any changes, which has no effect on the template).

Differences to proposal, general observations:
- currently, there is no place for a user to manage their templates. They can only add new ones, but not change existing ones. Only Django admins may manage templates, including making them publicly visible.
- saving a template with the same name as another (regardless of different user or not) is allowed. This does not change the other template. Templates are differentiated in the database by an integer ID. I added the proposed suffixes anyway to better differentiate templates.
- description is only shown in template dropdown selection as hover tooltip, creation date and user are not shown (but saved in DB)
- as of now, the checkbox "request to make this standard" is not doing much. This needs some internal procedure. How do we want to handle these requests?
- selected timeseries are saved and loaded as well, but plot does not refresh
- problem: timeseries are referenced by ID. In the DB, they are linked to a user and a scenario (not a project). When deleting either that user or that scenario, the timeseries is deleted as well, invalidating all derived assets. This seems to be a general issue with how timeseries are saved and handled.

Commit:
add AssetTemplate

paulapreuss commented on app/templates/asset/asset_create_form.html:
Is there a reason you are adding the template fields in the raw html instead of implementing it through a `ModelForm` from `AssetTemplate`? That would probably be easier to style given our current formats and there is less maintenance to be done if we end up changing any of the fields.

stefansc1 replied:
I wanted the template code to be explicit and readable and did not intend to use a Django form to send template data back when saving. If it were part of AssetCreateForm, the fields would be rendered alongside the other fields, so I would have needed some more custom filters. If it were not (making this its own form), it's even more complicated, since the HTML is replaced and part of another form. And I would have needed to send that other form data for each request.
I found it easier to just build the necessary fields myself and then simply posting data back via fetch. Building new FormData is required because we need to differentiate between asset data and template data (like template name and description). I am reluctant to send both in the same payload and splitting information server-side. Easier to just dump the asset data in the frontend and appending that as a neat little package to our post, so the backend does not need to actually do inspection and can just save to DB. Also, building FormData from scratch is nothing new, the same mechanic is used when saving assets.
Styling should actually be easier because we have direct access to all fields, without Django magic in-between.

paulapreuss commented on app/static/js/grid_model_topology.js:
```diff
@@ -747,3 +749,77 @@ function computeCOP(event){
+        if (typeof value === "string")
+            data[key] = value;
+    });
+    const templateFormData = new FormData();
```
Like here, for example, it seems like it would be cleaner to already have the form data instead of creating a new object here and populating it.

paulapreuss commented on app/templates/asset/asset_create_form.html:
```diff
@@ -97,5 +97,39 @@ <h3>{% translate "Technical parameters" %}</h3>
+          <h4>{% translate "Use existing" %}</h4>
+          <select id="template-select" style="width:100%">
+            <option value="">---</option>
+            {% for template in templates %}
```
When first adding an asset, my template drop down stays empty (although I have previously saved templates). Only when I close the form and open it again does the drop down get populated with my existing templates, for some reason.

stefansc1 replied:
Good catch. It has to do with the initial saving when adding a new asset. I assume you had an some field with an error (like missing timeseries)? In that case, the inner HTML was replaced to show these error messages, but the template data got lost. I added the template collection to the asset saving process. Also added templates to storage assets (they have their own forms).

paulapreuss commented on app/projects/views.py:
```diff
@@ -1845,6 +1874,38 @@ def asset_cops_create_or_update(
+        template = get_object_or_404(AssetTemplate, id=int(request.GET.get("id")))
+        # check permissions
+        if template.visibility == "project" and project_id != template.project_id:
+            raise Http404()
```
```suggestion
            raise PermissionDenied
```

stefansc1 replied:
Changed.

paulapreuss reviewed (commented):
Thank you for this! 🚀 
I think it will be a cool feature once it is done. I agree about the need to have some sort of management over the created template, I created a dedicated issue here #403. We can already use the changes in this PR to start testing on the staging server and collect feedback about how the feature should be further developed / how the handling should be for the things that are unclear. Once the functionality is clear, we can start to implement the design suggestions from #283.

Commit:
template fixes

- make templates available for storage assets
- show templates even if saving asset failed
